packages/ethereum-contracts/test/foundry/libs/SlotsBitmapLibrary.prop.sol

```diff
@@ -0,0 +1,204 @@
+        (uint32[] memory slotIds, ) = _listData(superToken, subscriber);
+        assertEq(slotIds.length, 0);
+        for (uint8 i = 0; i < numSlots; i++) {
+            _findEmptySlotAndFill(superToken, subscriber, fakeId);
+        }
+        (slotIds, ) = _listData(superToken, subscriber);
+        assertEq(slotIds.length, numSlots);
+
+        // remove an arbitary slot id between 0 and numSlots - 1
+        subIdToRemove = uint8(bound(uint256(subIdToRemove), 0, numSlots - 1));
+        _clearSlot(superToken, subscriber, subIdToRemove);
+        
+        (slotIds, ) = _listData(superToken, subscriber);
+        // the deleted slot should no longer exist
+        bool slotExists = _slotExists(slotIds, subIdToRemove);
+        assert(!slotExists);
+
+        // length should be numSlots - 1
+        assertEq(slotIds.length, numSlots - 1);
+    }
+
+    // randomize the sequence of the two instructions (findEmptySlotAndFill and clearSlot)
+    function testRandomSlotsBitmapLibraryBehavior(
+        uint8 numSlots,
+        int256[] memory instructions
+    ) public {
+        vm.assume(numSlots > 0);
+
+        // instructions is a random array of int256 numbers
+        for (int256 i = 0; i < int256(instructions.length); i++) {
+            (uint32[] memory slotIds, ) = _listData(superToken, subscriber);
+            uint256 currentLength = slotIds.length;
+            // find way to randomize this number
+            uint8 randomSlotIdToClear = uint8(uint256(instructions[uint256(i)]));
+            randomSlotIdToClear = uint8(bound(uint256(randomSlotIdToClear), 0, type(uint8).max));
+
+            // clearSlot when number is less than 0 AND there are slots to clear
+            if (instructions[uint256(i)] < 0 && slotIds.length > 0) {
+                bool slotExists = _slotExists(slotIds, randomSlotIdToClear);
+
+                // if the random slot we want to delete exists, we clear it
+                // and length should be one less now
+                if (slotExists) {
+                    _clearSlot(superToken, subscriber, randomSlotIdToClear);
+                    (slotIds, ) = _listData(superToken, subscriber);
+
+                    assertEq(slotIds.length, currentLength - 1);
+                    // the deleted slot should no longer exist
+                    slotExists = _slotExists(slotIds, randomSlotIdToClear);
+                    assert(!slotExists);
+                }
+            }
+
+            // findEmptySlotAndFill when number is greater than 0 AND the slotIds isn't maxed out
+            if (
+                instructions[uint256(i)] > 0 && slotIds.length < type(uint8).max
+            ) {
+                // we find an empty slot and fill it and length should be one more now
+                _findEmptySlotAndFill(superToken, subscriber, fakeId);
+                (slotIds, ) = _listData(superToken, subscriber);
+                assertEq(slotIds.length, currentLength + 1);
+            }
+        }
```

Comment:
should add an invariance check

Reply:
or put another way, what is the expected return result from _listData? I would refactor the test to check exactly what is the expected values. E.g. (pseudo code, not solidity) :

```
_listData(superToken, subscriber, [0, 1, 2, 4] /* expected slotIds */ )
```

Reply:
okay, from what I understand, this will require recreation of the expected slot ids array using the same or similar logic as listData though (without assembly) as the `listData` function filters out any "empty" slots.

is this ok?

Reply:
This is also a "property" of slotsbitmap in pseudo code:

```
l1 <- listSlots
foreach i  in l1: clearSlot i
l2 <- listSlots
assert l2 is empty
```

Does it make sense?

Reply:
yeah, makes sense

Reply:
added the property to test, are we sure we want to add that assert in `clearSlot` - our fuzzing doesn't break when adding it in.

Reply:
yes we do, it's better to asserts and burn a bit gas than letting bad cases survive in the wild.

Reply:
okay.

I would try a bit more:

1. make testRandomSlotsBitmapLibraryBehavior a shared function _createRandomSlots
2. combine _createRandomSlots + testCreateSlotsAndClearAll
3. a new test case: combine _createRandomSlots + createTooManySlots, and the last creation should revert

Not sure about the purpose of those non "property" single test cases , but no harm to keep them neither.

Reply:
```
    function testRandomSlotsBitmapLibraryBehavior(
        uint8 numSlots,
        int256[] memory instructions
    ) public {
 ```

numSlots is not used, instead I would do assume(instructions.length > 0)